### ISSUE TITLE
fs.readFile() can return string?

### ISSUE THREAD

**Issue by bolinfest** [facebook/flow#416]
It appears `fs.readFile()` can return a `string` instead of a `Buffer` if an encoding option is specified, such as:

    fs.readFile(json, {encoding: 'UTF-8'}, function(data) { JSON.parse(data); });

As such, I think the definition needs to be changed to:

```
  declare function readFile(
    filename: string,
    options?: Object | string,
    callback?: (err: ?Error, data: Buffer | string) => void
  ): void;
```

**Comment by gyzerok** [comment]
Are you sure that it is not an implicit type cast?

**Comment by bolinfest** [comment]
Yes: I checked with typeof.

**Comment by gyzerok** [comment]
@bolinfest ok, thanks. Suppose ```readFileSync``` behaves similarly?

**Comment by samwgoldman** [comment]
We talked about this problem on IRC, and the `Buffer | string` type doesn't really capture the actual type signature very elegantly. In particular, even if you know you provided an encoding or not, callers would still need to refine the type of data in order to use string-only or Buffer-only properties, so lots of existing, untyped code will throw errors.

Ideally, we could declare this function to properly express the dependency between the `options` type and the return type. Something like:

```
declare function readFile(
  filename: string,
  options: { encoding: string; flag?: string },
  callback: (err: ?Error, data: string) => void
)
declare function readFile(
  filename: string,
  options?: { flag?: string },
  callback: (err: ?Error, data: Buffer) => void
)
```

For now, I think the best course of action is to type `data` as `any`. This way, existing code will not cause type errors. We should spend time making overloaded declarations work as expected.

/cc @gabelevi

**Comment by impinball** [comment]
I would say the best fix is to allow for method signature overloading in method declarations.

**Comment by lapanoid** [comment]
+1 for fixing this bug, I guess that usage in this form fs.readFile(filename, 'utf-8') is even more popular than without encoding.

**Comment by impinball** [comment]
Eh... Apparently so.

**Comment by lumberj** [comment]
Is anyone actively working on this?

**Comment by samwgoldman** [comment]
@lumberj #653

**Comment by samwgoldman** [comment]
Fixed in 0.14
